packages/apollo-link-dedup/src/dedupLink.ts

```diff
@@ -16,16 +16,26 @@ export class DedupLink extends ApolloLink {
   > = new Map();
   private subscribers: Map<string, any> = new Map();
 
+  constructor({ useContext = false }: { useContext?: boolean } = {}) {
+    super();
+    this.useContextForDedup = useContext;
+  }
+
   public request(
     operation: Operation,
     forward: NextLink,
   ): Observable<FetchResult> {
+    const context = operation.getContext();
     // sometimes we might not want to deduplicate a request, for example when we want to force fetch it.
-    if (operation.getContext().forceFetch) {
+    if (context.forceFetch) {
       return forward(operation);
     }
 
-    const key = operation.toKey();
+    let key = operation.toKey();
+    if (this.useContextForDedup) {
+      //may throw error if config not serializable
+      key = `${key}-${JSON.stringify(context)}`;
```

Comment:
Similar pattern/idea from here: https://github.com/apollographql/apollo-link/blob/master/packages/apollo-link-batch-http/src/batchHttpLink.ts#L207

Reply:
Shouldn't this logic be moved into `operation.toKey`, so that we don't have multiple layers of key building?

Reply:
@benjamn I considered moving it there but I wasn't sure if that was the preferred approach since there was already some precedent for building the key in the specific package - e.g. the link above for `apollo-link-batch-http`. This approach also minimized the blast radius by being opt-in only.

Even though it feels like this should be the default behavior, if this is moved to `operation.toKey` with no opt-in mechanism, it technically becomes a breaking change. I suppose we could still check the context for some option, but that might seem a little strange as a consumer to specify on every request.

Would you rather it be slightly clunky for the sake of being opt-in or just have it be the default behavior?